File: App/MeasurementRecord.swift
import Foundation
import SwiftData
import TurntableCore

/// 一次量測的存檔。
///
/// **只存分析後的結論，不存逐樣本原始資料。** 一次 3 分鐘的量測是兩萬筆樣本、
/// 約 700 KB，存個幾十次資料庫就肥了。原始資料要留的話用「匯出」，那是刻意的
/// 一次性動作。歷史記錄要回答的是「上次調整之後有沒有變好」，摘要就夠。
@Model
final class MeasurementRecord {
    var date: Date = Date()
    var note: String = ""

    // 轉速
    var meanRPM: Double = 0
    var nominalLabel: String?
    var errorPercent: Double?
    /// 當下套用的校準倍率。沒校準是 nil —— 跟「倍率剛好是 1」要分得開。
    var calibrationFactor: Double?

    // 量測條件
    var durationSeconds: Double = 0
    var revolutions: Int = 0
    /// 被穩定閘門切掉的秒數。大於零代表操作或盤面有狀況。
    var trimmedSeconds: Double = 0

    // 分析
    var wrmsPercent: Double = 0
    var peak2SigmaPercent: Double = 0
    var onePerRevPercent: Double = 0
    var rotationHz: Double = 0
    var dominantPeakShare: Double = 0
    /// 譜峰用 JSON 存。SwiftData 對巢狀型別的支援綁手綁腳，而這是唯讀的展示資料，
    /// 不需要查詢或關聯。
    var peaksJSON: Data?

    init() {}

    /// 從一次分析建立記錄。
    convenience init(analysis: MeasurementAnalysis,
                     snapshot: MotionEngine.Snapshot,
                     calibration: Double?) {
        self.init()
        date = Date()
        meanRPM = analysis.meanRPM * (calibration ?? 1.0)
        nominalLabel = snapshot.nominal?.label
        errorPercent = snapshot.errorPercent
        calibrationFactor = calibration
        durationSeconds = analysis.durationSeconds
        revolutions = snapshot.revolutions
        trimmedSeconds = analysis.trimmedStartSeconds + analysis.trimmedEndSeconds
        wrmsPercent = analysis.wowFlutter.wrmsPercent
        peak2SigmaPercent = analysis.wowFlutter.peak2SigmaPercent
        onePerRevPercent = analysis.onePerRevolutionPercent
        rotationHz = analysis.rotationHz
        dominantPeakShare = analysis.dominantPeakShare
        peaksJSON = try? JSONEncoder().encode(
            analysis.peaks.prefix(6).map(StoredPeak.init))
    }

    var peaks: [StoredPeak] {
        guard let peaksJSON else { return [] }
        return (try? JSONDecoder().decode([StoredPeak].self, from: peaksJSON)) ?? []
    }

    var isCalibrated: Bool { calibrationFactor != nil }
}

/// 存檔用的譜峰。刻意跟 `SpectralPeak` 分開 —— 核心的型別可能會演進，
/// 存檔格式不該跟著動，否則舊記錄會讀不出來。
struct StoredPeak: Codable, Identifiable {
    var id: Double { frequencyHz }
    let frequencyHz: Double
    let amplitudePercent: Double
    let orderOfRotation: Double
    let isHarmonic: Bool

    init(_ p: SpectralPeak) {
        frequencyHz = p.frequencyHz
        amplitudePercent = p.amplitudePercent
        orderOfRotation = p.orderOfRotation
        isHarmonic = p.isRotationHarmonic
    }

    /// 判讀文字用存下來的欄位重建，不依賴核心的型別。
    var interpretation: String {
        let n = Int(orderOfRotation.rounded())
        if isHarmonic {
            switch n {
            case 1:  return "每圈一次 —— 偏心"
            case 2:  return "每圈兩次 —— 盤面橢圓"
            default: return "轉盤 \(n)× 諧波"
            }
        }
        if orderOfRotation < 1 { return "比一圈還慢 —— 皮帶循環" }
        return String(format: "非諧波（%.1f 倍）—— 傳動鏈", orderOfRotation)
    }
}

File: android/app/src/main/java/com/roger/turntablerpm/history/MeasurementRecord.kt
package com.roger.turntablerpm.history

import com.roger.turntablerpm.core.MeasurementAnalysis
import com.roger.turntablerpm.core.SpectralPeak
import org.json.JSONArray
import org.json.JSONObject

/**
 * 一次量測的存檔。
 *
 * **只存分析後的結論，不存逐樣本原始資料。** 一次 3 分鐘的量測是兩萬筆樣本、
 * 約 2 MB，存幾十次就爆了。歷史記錄要回答的是「上一次調整之後有沒有變好」，
 * 以及「換一個擺法之後 1× 有沒有變」—— 摘要就夠。
 *
 * **譜峰一定要存。** 沒有它就做不了「手機轉 180° 再量一次」這種需要比對
 * 個別成分的實驗，而那正是這個 app 最有價值的用法。
 */
data class MeasurementRecord(
    val epochMillis: Long,
    val note: String,
    val meanRPM: Double,
    val rawMeanRPM: Double,
    val calibrationFactor: Double?,
    val nominalLabel: String?,
    val errorPercent: Double?,
    val durationSeconds: Double,
    /** 這段量測轉了幾圈。iOS 端有存，移植時漏掉了 —— 詳情頁要顯示。 */
    val revolutions: Int,
    val trimmedSeconds: Double,
    val sampleRate: Double,
    val rotationHz: Double,
    val wrmsPercent: Double,
    val peak2SigmaPercent: Double,
    val onePerRevPercent: Double,
    val dominantPeakShare: Double,
    val peakAngleDegrees: Double?,
    val peaks: List<StoredPeak>,
) {
    val isCalibrated: Boolean get() = calibrationFactor != null

    fun toJson(): JSONObject = JSONObject().apply {
        put("t", epochMillis); put("note", note)
        put("mean", meanRPM); put("raw", rawMeanRPM)
        calibrationFactor?.let { put("k", it) }
        nominalLabel?.let { put("nom", it) }
        errorPercent?.let { put("err", it) }
        put("dur", durationSeconds); put("rev", revolutions)
        put("trim", trimmedSeconds); put("fs", sampleRate)
        put("rot", rotationHz); put("wrms", wrmsPercent); put("p2s", peak2SigmaPercent)
        put("one", onePerRevPercent); put("share", dominantPeakShare)
        peakAngleDegrees?.let { put("angle", it) }
        put("peaks", JSONArray().apply { peaks.forEach { put(it.toJson()) } })
    }

    companion object {
        fun from(
            analysis: MeasurementAnalysis,
            rawMeanRPM: Double,
            revolutions: Int,
            calibrationFactor: Double?,
            nominalLabel: String?,
            errorPercent: Double?,
            note: String = "",
            epochMillis: Long = System.currentTimeMillis(),
        ) = MeasurementRecord(
            epochMillis = epochMillis,
            note = note,
            meanRPM = analysis.meanRPM,
            rawMeanRPM = rawMeanRPM,
            calibrationFactor = calibrationFactor,
            nominalLabel = nominalLabel,
            errorPercent = errorPercent,
            durationSeconds = analysis.durationSeconds,
            revolutions = revolutions,
            trimmedSeconds = analysis.trimmedStartSeconds + analysis.trimmedEndSeconds,
            sampleRate = analysis.sampleRate,
            rotationHz = analysis.rotationHz,
            wrmsPercent = analysis.wowFlutter.wrmsPercent,
            peak2SigmaPercent = analysis.wowFlutter.peak2SigmaPercent,
            onePerRevPercent = analysis.onePerRevolutionPercent,
            dominantPeakShare = analysis.dominantPeakShare,
            peakAngleDegrees = analysis.peakAngleDegrees,
            peaks = analysis.peaks.take(8).map { StoredPeak.from(it) },
        )

        fun fromJson(o: JSONObject): MeasurementRecord {
            val arr = o.optJSONArray("peaks") ?: JSONArray()
            return MeasurementRecord(
                epochMillis = o.getLong("t"),
                note = o.optString("note", ""),
                meanRPM = o.getDouble("mean"),
                rawMeanRPM = o.optDouble("raw", o.getDouble("mean")),
                calibrationFactor = if (o.has("k")) o.getDouble("k") else null,
                nominalLabel = if (o.has("nom")) o.getString("nom") else null,
                errorPercent = if (o.has("err")) o.getDouble("err") else null,
                durationSeconds = o.getDouble("dur"),
                revolutions = o.optInt("rev", 0),
                trimmedSeconds = o.optDouble("trim", 0.0),
                sampleRate = o.optDouble("fs", 100.0),
                rotationHz = o.optDouble("rot", 0.0),
                wrmsPercent = o.getDouble("wrms"),
                peak2SigmaPercent = o.optDouble("p2s", 0.0),
                onePerRevPercent = o.optDouble("one", 0.0),
                dominantPeakShare = o.optDouble("share", 0.0),
                peakAngleDegrees = if (o.has("angle")) o.getDouble("angle") else null,
                peaks = (0 until arr.length()).map { StoredPeak.fromJson(arr.getJSONObject(it)) },
            )
        }
    }
}

/**
 * 存檔用的譜峰。**刻意跟核心的 `SpectralPeak` 分開** —— 核心的型別可能會演進，
 * 存檔格式不該跟著動，否則舊記錄會讀不出來。
 */
data class StoredPeak(
    val frequencyHz: Double,
    val amplitudePercent: Double,
    val orderOfRotation: Double,
    val isHarmonic: Boolean,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("hz", frequencyHz); put("amp", amplitudePercent)
        put("ord", orderOfRotation); put("harm", isHarmonic)
    }

    companion object {
        fun from(p: SpectralPeak) =
            StoredPeak(p.frequencyHz, p.amplitudePercent, p.orderOfRotation, p.isRotationHarmonic)

        fun fromJson(o: JSONObject) = StoredPeak(
            o.getDouble("hz"), o.getDouble("amp"), o.getDouble("ord"), o.getBoolean("harm"),
        )
    }
}
